apache/beam
# [BEAM-11105] Basic Watermark Estimation (Wall Clock Observing)

**Summary of Changes**

This pr is the first step towards full watermark estimation support - it introduces the mechanism used for watermark estimation including:

- Validation (graph/fn.go)
- Code generation (genx/genx.go)
- Execution (exec/sdf.go and exec/sdf_invokers.go)
- Propogating result back to the runner (exec/datasource.go, harness.go)
- A concrete implementation of a wall clock observing estimator (sdf/watermarkEstimator.go)

This is built according to the design specified here - https://docs.google.com/document/d/1DqCYJ-J1YGNelCRIcN5v6BQsZxJB2l5uWBmN4ti--Ew/edit?usp=sharing

**Next Steps**

This work does not attempt any of the following (though it is built in such a way that those should be cleanly supported in subsequent PRs):

- Watermark state
- Manual estimators (passed into ProcessElement)
- Timestamp observing estimators (call ObserveTimestamp on each emitted timestamp)

**Manual Testing**

On top of all the unit tests introduced, I hacked the existing textio sdf doFn and added the following function:

```
func (fn *readSdfFn) CreateWatermarkEstimator() sdf.WallTimeWatermarkEstimator {
	return sdf.WallTimeWatermarkEstimator{}
}
```

and ran wordCount (with the textio sdf version) on dataflow so that I could observe the output watermark was getting correctly updated to the current wall clock time:

<img width="464" alt="image" src="https://user-images.githubusercontent.com/42773683/161610141-f38ce9f0-a598-45bd-aa5b-869c465300a5.png">

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:


See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.
